### PR TITLE
Add SlicingClassifier

### PR DESCRIPTION
**Pull request by vincentschen**
**Changes**
* Add `SlicingBinaryClassifier` which subclasses `BaseMultitaskClassifier` and adds simple wrappers for scoring with slices

**Test Plan**
* Added class tests
* `tox` passes

### CODE REVIEW

File: snorkel/slicing/slicing_classifier.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+
+class BinarySlicingClassifier(MultitaskClassifier):
+    """A slice-aware binary classifier that supports training + scoring on slice labels.
+
+    Parameters
+    ----------
+    base_architecture
+        A network architecture that accepts input data and outputs a representation
+    head_dim
+        Output feature dimension of the representation_net, and input dimension of the
```
Comment (henryre) [inline snorkel/slicing/slicing_classifier.py]: repr_net to base_arch. use backticks to indicate it's a specific reference

File: snorkel/slicing/slicing_classifier.py [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+        self, dataset: DictDataset, S: np.ndarray, **dataloader_kwargs: Any
+    ) -> DictDataLoader:
+        """Create DictDataLoader with slice labels, initialized from specified dataset.
+
+        Parameters
+        ----------
+        dataset
+            A DictDataset that will be converted into a slice-aware dataloader
+        S
+            A [num_examples by num_slices] slice matrix indicating whether
```
Comment (henryre) [inline snorkel/slicing/slicing_classifier.py]: "by" -> x

File: snorkel/slicing/slicing_classifier.py [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+        # Make dataloader slice-aware
+        add_slice_labels(dataloader, self.base_task, S, self.slice_names)
+
+        return dataloader
+
+    @torch.no_grad()
+    def score_slices(
+        self, dataloaders: List[DictDataLoader], as_dataframe: bool = False
+    ) -> Dict[str, float]:
+        """Scores appropriate slice labels using the base_task (NOT slice_tasks).
```
Comment (henryre) [inline snorkel/slicing/slicing_classifier.py]: use backticks

File: snorkel/slicing/slicing_classifier.py [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+
+        # Make dataloader slice-aware
+        add_slice_labels(dataloader, self.base_task, S, self.slice_names)
+
+        return dataloader
+
+    @torch.no_grad()
+    def score_slices(
+        self, dataloaders: List[DictDataLoader], as_dataframe: bool = False
+    ) -> Dict[str, float]:
```
Comment (henryre) [inline snorkel/slicing/slicing_classifier.py]: return type should be union with pd.DataFrame. Same with score()

File: test/slicing/test_slicing_classifier.py [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+def g(x) -> int:
+    return x.num > 10
+
+
+DATA = [3, 43, 12, 9, 3]
+
+
+class SliceCombinerTest(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls):
```
Comment (henryre) [inline test/slicing/test_slicing_classifier.py]: probably just want setUp if we're calling train

File: test/slicing/test_slicing_classifier.py [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+        # No 'ind' labels!
+        self.assertNotIn(
+            "test_task_slice:hello_ind/test_dataset/valid/f1", slice_scores
+        )
+        self.assertNotIn(
+            "test_task_slice:world_ind/test_dataset/valid/f1", slice_scores
+        )
+
+
+def create_dataset(X, Y, split, dataset_name, input_name, task_name):
```
Comment (henryre) [inline test/slicing/test_slicing_classifier.py]: put before testcase

File: test/slicing/test_slicing_classifier.py [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+        """Ensure that the appropriate scores are returned with .score and .score_slices."""
+        # Make dataloaders
+        dataloaders = [
+            self.slice_model.make_slice_dataloader(dataset=ds, S=self.S, batch_size=4)
+            for ds in self.datasets
+        ]
+        train_dl, valid_dl = tuple(dataloaders)
+
+        # Train model
+        trainer = Trainer(n_epochs=1)
```
Comment (henryre) [inline test/slicing/test_slicing_classifier.py]: do we need to train first?

File: test/slicing/test_slicing_classifier.py [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+        trainer.fit(self.slice_model, [train_dl])
+
+        # Eval overall
```
Comment (henryre) [inline test/slicing/test_slicing_classifier.py]: great!

File: test/slicing/test_slicing_classifier.py [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+        dataset_name = "test_dataset"
+        splits = ["train", "valid"]
+        cls.datasets = [
+            create_dataset(
+                cls.X, cls.Y, split, dataset_name, cls.data_name, cls.task_name
+            )
+            for split in splits
+        ]
+
+        cls.slice_model = BinarySlicingClassifier(
```
Comment (henryre) [inline test/slicing/test_slicing_classifier.py]: this could probably use an end-to-end test, but let's punt for now

File: snorkel/slicing/slicing_classifier.py [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+
+        Parameters
+        ----------
+        dataset
+            A DictDataset that will be converted into a slice-aware dataloader
+        S
+            A [num_examples by num_slices] slice matrix indicating whether
+            each example is in every slice
+        slice_names
+            A list of slice names corresponding to columsn of ``S``
```
Comment (bhancock8) [inline snorkel/slicing/slicing_classifier.py]: columns

File: snorkel/slicing/slicing_classifier.py [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+    @torch.no_grad()
+    def score_slices(
+        self, dataloaders: List[DictDataLoader], as_dataframe: bool = False
+    ) -> Dict[str, float]:
+        """Scores appropriate slice labels using the base_task (NOT slice_tasks).
+
+        In practice, we'd like to use a final prediction from a _single_ task head.
+        To do so, ``self.base_task`` leverages reweighted slice representation to
+        make a prediction. In this method, we remap all slice-specific ``pred``
+        labels to ``self.base_task`` for evaluation.
```
Comment (bhancock8) [inline snorkel/slicing/slicing_classifier.py]: In other words, this method will tell us how well our overall prediction head performs on individual slices, correct?

File: snorkel/slicing/slicing_classifier.py (hunk 2 above)
Comment (bhancock8) [reply snorkel/slicing/slicing_classifier.py]: or [num_examples, num_slices]

File: snorkel/slicing/slicing_classifier.py (hunk 1 above)
Comment (vincentschen) [reply snorkel/slicing/slicing_classifier.py]: 👍

File: test/slicing/test_slicing_classifier.py (hunk 7 above)
Comment (vincentschen) [reply test/slicing/test_slicing_classifier.py]: nope, removed

File: snorkel/slicing/slicing_classifier.py [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+
+    Parameters
+    ----------
+    base_architecture
+        A network architecture that accepts input data and outputs a representation
+    head_dim
+        Output feature dimension of the representation_net, and input dimension of the
+        internal prediction head: ``nn.Linear(head_dim, 2)``.
+    slice_names
+        A list of slice names that the model will accept as tasks/labels
```
[the remark this answers is not in the document]
Comment (vincentschen) [reply snorkel/slicing/slicing_classifier.py]: >         A list of slice names that the model will accept initialize as tasks
        and accept as corresponding labels

File: snorkel/slicing/slicing_classifier.py (hunk 11 above)
Comment (vincentschen) [reply snorkel/slicing/slicing_classifier.py]: yes! used higher-level language in comments
